File: logger1.cpp
#include <iostream>
#include <thread>
#include <vector>

int main() {
  std::vector<int> logger;

  std::vector<std::thread> handlers;
  for(int i = 0; i<10; i++) {
    handlers.push_back(std::thread([i, &logger] {
      // XXX its invalid!
      logger.push_back(i);
    }));
  }

  for(auto& h : handlers) {
    h.join();
  }

  for(auto log : logger) {
    std::cout << log << ", ";
  }
  std::cout << std::endl;

  return 0;
}

File: logger1.rs
use std::io::prelude::*;
use std::thread;

fn main() {
    let mut logger = vec![];

    let mut handlers = vec![];
    for i in 0..10 {
        handlers.push(
            thread::spawn(move || {
                // ... do something ...

                // NOTE cannot compile because here:
                logger.push(i);
            }
        ));
    }

    for h in handlers {
        h.join().unwrap();
    }
}
